src/daemon/daemon.cpp

```diff
@@ -1795,19 +1923,56 @@ QJsonObject mp::Daemon::retrieve_metadata_for(const std::string& name)
     return vm_instance_specs[name].metadata;
 }
 
+std::string mp::Daemon::generate_unused_mac_address()
+{
+    std::string mac_address = mp::utils::generate_mac_address();
+
+    // TODO: Checking in our list of MAC addresses does not suffice to conclude the generated MAC is unique. We
+    // should also check in the ARP table.
+    while (allocated_mac_addrs.find(mac_address) != allocated_mac_addrs.end())
+    {
+        mac_address = mp::utils::generate_mac_address();
+    }
+
+    allocated_mac_addrs.insert(mac_address);
+
+    return mac_address;
+}
+
 void mp::Daemon::persist_instances()
 {
     auto vm_spec_to_json = [](const mp::VMSpecs& specs) -> QJsonObject {
         QJsonObject json;
         json.insert("num_cores", specs.num_cores);
         json.insert("mem_size", QString::number(specs.mem_size.in_bytes()));
         json.insert("disk_space", QString::number(specs.disk_space.in_bytes()));
-        json.insert("mac_addr", QString::fromStdString(specs.mac_addr));
         json.insert("ssh_username", QString::fromStdString(specs.ssh_username));
         json.insert("state", static_cast<int>(specs.state));
         json.insert("deleted", specs.deleted);
         json.insert("metadata", specs.metadata);
 
+        if (0 == specs.interfaces.size())
```

Comment:
```suggestion
        if (specs.interfaces.empty())
```

Reply:
I separated the default network interface from the list of extra interfaces. This is resolved.